flake.nix

```diff
@@ -0,0 +1,67 @@
+{
+  description = "tyrian-dev";
+
+  inputs = {
+    nixpkgs.url = github:nixos/nixpkgs/nixpkgs-unstable;
+    flake-utils.url = github:numtide/flake-utils;
+    flake-compat = {
+      url = github:edolstra/flake-compat;
+      flake = false;
+    };
+  };
+
+  outputs = { self, nixpkgs, flake-utils, ... }:
+    let
+      supportedSystems = [ "aarch64-darwin" "aarch64-linux" "x86_64-linux" ];
+
+      forSystem = system:
+        let
+          mill-overlay = f: p: {
+            # top-level/all-packages.nix hardcodes `jre=jre8` when building mill so overriding jre alone does not work
+            mill = p.mill.overrideAttrs (old: {
+              installPhase = ''
+                runHook preInstall
+                install -Dm555 "$src" "$out/bin/.mill-wrapped"
+                # can't use wrapProgram because it sets --argv0
+                makeWrapper "$out/bin/.mill-wrapped" "$out/bin/mill" \
+                  --prefix PATH : "${p.jdk17_headless}/bin" \
+                  --set JAVA_HOME "${p.jdk17_headless}"
+                runHook postInstall
+              '';
+            });
+          };
+
+          pkgs = import nixpkgs {
+            inherit system;
+            overlays = [ mill-overlay ];
+          };
+          jdk = pkgs.jdk17_headless;
+        in
+        {
+          devShells = {
+            jvm = pkgs.mkShell {
+              name = "scala-dev-shell";
+              buildInputs = [
+                jdk
+                pkgs.coursier
+                pkgs.mill
+                pkgs.sbt
+              ];
+              shellHook = ''
+                JAVA_HOME="${jdk}"
+              '';
+            };
+
+            js = pkgs.mkShell {
+              name = "js-dev-shell";
+              buildInputs = with pkgs; [
+                nodejs
+                yarn
```

Comment:
I think that's it, otherwise you end up doing things like pinning npm packages which should probably be avoided?

Nix-noob question: Is it a good idea to run `yarn install` when the environment is switched to so make sure that you can, for instance, run the tests? Or is that sort of practice to be avoided?

Reply:
Yes, pinning npm packages is left to node to handle (or there are other mode advanced tools like `node2nix` but I wouldn't go that far for a nix shell).

I do have a `shellHook` in my application shell that runs both `yarn install` and `yarn build`, but I'm not sure if this would be a good idea in Tyrian, as there are multiple `yarn` projects in different folders.

Reply:
Done 👍🏽

Reply:
I thought about it a bit, since you mentioned `sbt test` together with `yarn install`, maybe it makes sense to have a single shell with all `jvm` and `js` software?

Reply:
Now you mention it, that does seem like a good idea. I was going to give it a try later and see how it all felt, but I'm thrilled to take your advice.